pkg/kubernetes/deploy.js

```diff
@@ -0,0 +1,373 @@
+                    is_deploying.parent().prepend($(Mustache.render(deploy_notification_success)));
+
+                })
+                .fail(function(ex, response) {
+                    /* ex containst the failure */
+                    var jdata = JSON.parse(response);
+                    hide_progress_message();
+                    var context = {};
+                    is_deploying.parent().prepend($(Mustache.render(deploy_notification_failure, $.extend(context, jdata))));
+
+                });
+        }
+
+        function create_everything(namespace, services, rcs, pods) {
+            var ns_json = '{"apiVersion":"v1beta3","kind":"Namespace","metadata":{"name": "' + namespace + '"}}';
+            var tasks = [];
+
+            tasks.push(["namespace", [ns_json], client.create_ns]);
+
+            for (var serv in services) {
+                tasks.push([services[serv].metadata.name, [namespace, JSON.stringify(services[serv])], client.create_service]);
+            }
+
+            for (var rc in rcs) {
+                tasks.push([rcs[rc].metadata.name, [namespace, JSON.stringify(rcs[rc])], client.create_replicationcontroller]);
+            }
+
+            for (var p in pods) {
+                tasks.push([pods[p].metadata.name, [namespace, JSON.stringify(pods[p])], client.create_pod]);
+            }
+
+            var deferred = $.Deferred();
+            var total = tasks.length;
+
+            function step(qtasks) {
+                var e = qtasks.shift();
+                if (!e) {
+                    deferred.resolve();
+                    return;
+                }
+                var task_name = e[0];
+                var args = e[1];
+                var task = e[2];
+
+                task.apply(null, args).done(function(response) {
+                    deferred.notify("created", response, qtasks.length - total, total);
+                    step(qtasks);
+                }).fail(function(ex, response) {
+                    if (ex.status == 409) {
+                        deferred.notify("skipped", response, qtasks.length - total, total);
+                        step(qtasks);
+                    } else {
+                        deferred.reject(ex, response);
+                    }
+                });
+            }
+
+            step(tasks);
+
+            return deferred.promise();
+        }
+
+        function show_progress_message(message) {
+            is_deploying.show();
+            $('#deploy-update-message').text(message);
+        }
+
+        function hide_progress_message() {
+            is_deploying.hide();
+        }
+
+        function remove_notifications() {
+            $('div.container-fluid.alert').remove();
+            deploy_dialog_remove_errors();
+        }
+
+
+        function action_in_progress() {
+            return (is_deploying.is(':visible'));
+        }
+
+        return {
+            'deploy_app': deploy_app
+        };
+
+    }
+
+    function deploy_app() {
```

Comment:
There are two functions in this file called deplop_app() albeit in different scopes. This is a bit confusing, Should we rename one?